libstb/secvar/backend/edk2-compat-util.c

```diff
@@ -0,0 +1,601 @@
+/* Copyright 2019 IBM Corp. */
+#ifndef pr_fmt
+#define pr_fmt(fmt) "EDK2_COMPAT: " fmt
+#endif
+
+#include <opal.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+#include <stdint.h>
+#include <ccan/endian/endian.h>
+#include <mbedtls/error.h>
+#include <device.h>
+#include "libstb/crypto/pkcs7/pkcs7.h"
+#include "edk2.h"
+#include "../secvar.h"
+#include "../secvar_devtree.h"
+#include "edk2-compat-util.h"
+
+char *utf8_to_ucs2(const char *key, size_t keylen)
+{
+	int i;
+	char *str;
+
+	str = zalloc(keylen * 2);
+	if (!str)
+		return NULL;
+
+	for (i = 0; i < keylen*2; key++) {
+		str[i++] = *key;
+		str[i++] = '\0';
+	}
+
+	return str;
+}
+
+bool key_equals(const char *key1, const char *key2)
+{
+	if (!strncmp(key1, key2, strlen(key1)))
+		return true;
+
+	return false;
+}
+
+void get_key_authority(const char *ret[3], const char *key)
+{
+	int i = 0;
+
+	if (key_equals(key, "PK"))
+		ret[i++] = "PK";
+	if (key_equals(key, "KEK"))
+		ret[i++] = "PK";
+	if (key_equals(key, "db") || key_equals(key, "dbx")) {
+		ret[i++] = "KEK";
+		ret[i++] = "PK";
+	}
+	ret[i] = NULL;
+}
+
+int get_esl_signature_list_size(char *buf, size_t buflen)
+{
+	EFI_SIGNATURE_LIST list;
+
+	if (buflen < sizeof(EFI_SIGNATURE_LIST))
+		return OPAL_PARAMETER;
+
+	memcpy(&list, buf, sizeof(EFI_SIGNATURE_LIST));
+
+	prlog(PR_DEBUG, "size of signature list size is %u\n", le32_to_cpu(list.SignatureListSize));
+
+	return le32_to_cpu(list.SignatureListSize);
+}
+
+int get_esl_cert(char *buf, size_t buflen, char **cert)
+{
+	size_t sig_data_offset;
+	size_t size;
+	EFI_SIGNATURE_LIST list;
+
+	if (buflen < sizeof(EFI_SIGNATURE_LIST))
+		return OPAL_PARAMETER;
+
+	memcpy(&list, buf, sizeof(EFI_SIGNATURE_LIST));
+
+	size = le32_to_cpu(list.SignatureSize) - sizeof(uuid_t);
+	/* No certificate in the ESL */
+	if (size <= 0)
+		return OPAL_PERMISSION;
+
+	if (!cert)
+		return OPAL_PARAMETER;
+
+	*cert = zalloc(size);
+	if (!(*cert))
+		return OPAL_NO_MEM;
+
+	prlog(PR_DEBUG,"size of signature list size is %u\n", le32_to_cpu(list.SignatureListSize));
+	prlog(PR_DEBUG, "size of signature header size is %u\n", le32_to_cpu(list.SignatureHeaderSize));
+	prlog(PR_DEBUG, "size of signature size is %u\n", le32_to_cpu(list.SignatureSize));
+
+	sig_data_offset = sizeof(list) + le32_to_cpu(list.SignatureHeaderSize) + 16 * sizeof(uint8_t);
+	if (sig_data_offset > buflen) {
+		free(*cert);
+		return OPAL_PARAMETER;
+	}
+	
+	memcpy(*cert, buf + sig_data_offset, size);
+
+	return size;
+}
+
+int get_pkcs7_len(struct efi_variable_authentication_2 *auth)
+{
+	uint32_t dw_length; 
+	size_t size;
+
+	if (!auth)
+		return OPAL_PARAMETER;
+	
+	dw_length = le32_to_cpu(auth->auth_info.hdr.dw_length);
+	size = dw_length - (sizeof(auth->auth_info.hdr.dw_length)
+			+ sizeof(auth->auth_info.hdr.w_revision)
+			+ sizeof(auth->auth_info.hdr.w_certificate_type)
+			+ sizeof(auth->auth_info.cert_type));
+
+	return size;
+}
+
+int get_auth_descriptor2(void *buf, size_t buflen, char **auth_buffer)
```

Comment:
This function definitely needs a comment on why it is an `auth_descriptor2`, and where `1` is, so someone less familiar with the format understands